control/controlclient/direct.go

```diff
@@ -703,6 +712,16 @@ func (c *Direct) sendMapRequest(ctx context.Context, maxPolls int, cb func(*netm
 		Hostinfo:      hi,
 		DebugFlags:    c.debugFlags,
 		OmitPeers:     cb == nil,
+
+		// On initial startup before we know our endpoints, set the ReadOnly flag
+		// to tell the control server not to distribute out our (empty) endpoints to peers.
+		// Presumably we'll learn our endpoints in a half second and do another post
+		// with useful results. The first POST just gets us the DERP map which we
+		// need to do the STUN queries to discover our endpoints.
+		// TODO(bradfitz): we skip this optimization in tests, though,
+		// because the e2e tests are currently hyperspecific about the
+		// ordering of things. The e2e tests need love.
+		ReadOnly: readOnly || (len(epStrs) == 0 && !everEndpoints && !inTest()),
```

Comment:
Something I observed in testing is endpoints being sent to the control server before a DERP connection had been established. In fact, the entire NetInfo block of Hostinfo was nil.

I don't know how common this is in reality, but maybe we should also set ReadOnly if `hi.NetInfo == nil`?

Reply:
This change has no change in behavior. This is just refactoring.

The plan is for PollNetMap to always be read-only and stream forever without ever being interrupted and any changes to netmap/etc will happen in lite endpoint update requests that don't abort the long poll.